### PR TITLE
Cleanup names of debug stats and trace contexts.

### PR DESCRIPTION
**Pull request by AndrewMeadows**
- cleanup timing stats names
- don't show Bullet simulation timing stats when disabling this menu option: **Developer --> Timing --> Performance Timer --> Expand Physics**
- add `Test.saveNextPhysicsStats()` to JS for saving detailed Bullet timings to file

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5585/HighFidelity-Beta-PR11965-74a1f5de9643bb06f0adec7b5dd437dfc1603eb3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5585/HighFidelity-Beta-PR11965-74a1f5de9643bb06f0adec7b5dd437dfc1603eb3.exe

**Comment by AndrewMeadows** [comment]
**Test plan:**

**(1)** Login using desktop mode (this test is easier without HMD).

**(2)** Enable this menu option: *Developer --> Timing --> Performance Timer --> Display Timing Details*

**(3)** Make sure this option is disabled: *Developer --> Timing --> Performance Timer --> Expand /physics*

**(4)** Show stats by pressing the *\* key.
--> You should see some transparent debug stats show up at the top-center of the screen.

**(5)** Click somewhere on the stats using the mouse to expand details.
--> You should see some runtime performance stats updating.

**(6)** With the stats showing enable this menu option: *Developer --> Timing --> Performance Timer --> Expand /physics*
--> On `hifi/master` You should see some stats that look something like: *.../some/BulletPhysics/Context*
--> On this PR you should see some stats that look something like */step/.../.../.../something*

**(7)** Disable this menu option: *Developer --> Timing --> Performance Timer --> Expand /physics*
--> On `hifi/master` you probably still see stats like *.../some/BulletPhysics/Context*
--> On this PR the bullet stats should vanish

The rest of the changes in this PR have to do with pure debug stuff: name changes for "trace" contexts, and the ability to write one frame's worth of Bullet timing measurements to file.  These capabilities don't need to be tested but for anyone interested here are some instructions on how to do it:

**(7)** Make a script that looks like this and save it to local file trace.js:
```
var startTime = 20;
var writeBulletStatsTime = startTime - 1;
var endTime = startTime + 5;
var bulletStatsFile = "bulletStats.txt";
var traceFile = "trace.json.gz";
Window.location = "hifi://dev-welcome";

Script.setTimeout(function() {
    Test.savePhysicsSimulationStats(bulletStatsFile);
}, startTime * 1000);

Script.setTimeout(function() {
    var loggingRules = "trace.*=true\n";
    Test.startTracing(loggingRules);
}, startTime * 1000);

Script.setTimeout(function() {
    Test.stopTracing(traceFile);
    Test.quit();
}, endTime * 1000);
```
**(8)** Launch the interface from the command line with arguments:

`interface.exe --testScript foo.js`

**(9)** Wait around for the script to run its course: it should automatically quit the interface after about 25 seconds up.

**(10)** Look in the current directory and you should see two new files:
```
bulletStats.txt
trace.json.gz
```
**(11)** The `bulletStats.txt` file is just a printout of some detailed stats in one Bullet step: it is human readable.

**(12)** The dats inside `trace.json.gz` can be viewed in the chromebrowser by navigating to:

chrome://tracing

and then loading the `trace.json.gz` file into the tool.  You can examine timing events in detail using tracing tool UI.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5601/HighFidelity-Beta-PR11965-87fa919ac8d2c38198c515df3e9dcdb49c32bc23.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5601/HighFidelity-Beta-PR11965-87fa919ac8d2c38198c515df3e9dcdb49c32bc23.dmg

**Comment by cain-kilgore** [comment]
Testing.

**Comment by cain-kilgore** [comment]
# Test Results

Test | Result
--- | ---
(1) Login using desktop mode (this test is easier without HMD). | Pass
(2) Enable this menu option: Developer --> Timing --> Performance Timer --> Display Timing Details | Pass
(3) Make sure this option is disabled: Developer --> Timing --> Performance Timer --> Expand /physics | Pass
(4) Show stats by pressing the / key. | Pass
--> You should see some transparent debug stats show up at the top-center of the screen. | Pass
(5) Click somewhere on the stats using the mouse to expand details. | Pass
--> You should see some runtime performance stats updating. | Pass
(6) With the stats showing enable this menu option: Developer --> Timing --> Performance Timer --> Expand /physics | Pass
--> On hifi/master You should see some stats that look something like: .../some/BulletPhysics/Context | Pass
--> On this PR you should see some stats that look something like /step/.../.../.../something | Pass
(7) Disable this menu option: Developer --> Timing --> Performance Timer --> Expand /physics | Pass
--> On hifi/master you probably still see stats like .../some/BulletPhysics/Context | Pass
--> On this PR the bullet stats should vanish | Pass
(8) Make a script that looks like this and save it to local file trace.js  | Pass
(9) Launch the interface from the command line with arguments | Pass
(10) Wait around for the script to run its course: it should automatically quit the interface after about 25 seconds up. | Pass
(11) Look in your Documents directory and you should see two new files: | Fail (bulletStats.txt not found)
(12) The bulletStats.txt file is just a printout of some detailed stats in one Bullet step: it is human readable. | Fail
(13) The data inside trace.json.gz can be viewed in the chromebrowser by navigating to chrome://tracing | Pass
and then loading the trace.json.gz file into the tool. You can examine timing events in detail using tracing tool UI. | Pass

image

I wasn't sure what result to give this as the bulletStats.txt file was not found during testing. If this isn't an issue then it should be a pass but if it is then this should be a fail. Can provide logs if needed.

**Comment by AndrewMeadows** [comment]
@cain-kilgore, sure send me the logs from your interface client session.  If the `bulletStats.txt` file could not be written it may have made some logs about it.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -4848,8 +4846,7 @@ void Application::update(float deltaTime) {
         if (_physicsEnabled) {
             {
                 PROFILE_RANGE(simulation_physics, "PreStep");
-
-                PerformanceTimer perfTimer("updateStates)");
+                PerformanceTimer perfTimer("preStep)");
```
Comment (NissimHadar) [inline interface/src/Application.cpp]: Is there a reason these are in separate blocks?

Comment (AndrewMeadows) [reply interface/src/Application.cpp]: Yes the block limits the scope of the `PerformanceTimer`.  Also, the `PROFILE_RANGE` macro creates a variable with a fixed name, which means two `PROFILE_RANGE` calls cannot be used in the exact same scope.

**Comment by AndrewMeadows** [comment]
This is actually a PASS on QA I think.  Both Nissim and I have tested the writing of the stats file and found it working.  I've tested on both Linux and Windows.  It's a debug feature anyway AND it logs when it fails to write to file (for permissions issues or bad path name).

File: interface/src/Application.cpp (hunk 1 above)
Comment (NissimHadar) [reply interface/src/Application.cpp]: Thanks!

**Pull request merged**